### PR TITLE
Add article relevance scoring and chunked articleRelevances

### LINKED ISSUE #216: [hyper-pm] MP-ART-ANALYSIS-006: Score relevance with canonical breakdown, version field, configurable selection, and chunked articleRelevances POST
**Ticket ID:** `MP-ART-ANALYSIS-006`  
**Priority:** P0  

---

# Score relevance with canonical breakdown, version field, configurable selection, and chunked articleRelevances POST

**Group:** Article Analysis Agent (`article-analysis-agent`)

## Summary

Computes per-article `score` in zero to one inclusive with a non-empty `scoreBreakdown` map. Uses PRD canonical v1 keys `breakingNews`, `kgRelation`, `fundamental`, `tickerSalience`, and `sourceQuality`, allows additional experimental numeric keys, and includes `_version` as an integer matching configurable `ARTICLE_ANALYSIS_SCORE_BREAKDOWN_VERSION` (MP-ART-ANALYSIS-009). Applies documented selection rules so `selected` reflects configurable thresholds and caps for downstream pipelines. POSTs `articleRelevances` in validated chunks per FR9.

## Scope

- In scope: Heuristic and or LLM-assisted relevance aligned with product priority dimensions breaking news, KG relation changes, and fundamentals as breakdown components (FR5). Selection logic configurable (threshold, max selected per ticker per day, optional diversity or intent mix caps as in PRD Hermes section). Chunked POST with schema validation. Align PRD breakdown semantics with observability canonical keys.
- Out of scope: Admin UI for tuning (use env or Hermes config). Optional `dedupOrOverlap` penalty can be stubbed or implemented if time permits.

## Acceptance criteria

- [x] Every relevance row has `score` in zero to one and non-empty `scoreBreakdown`.
- [ ] `scoreBreakdown` includes `_version` integer and all five canonical v1 keys as numeric components for core charts (additional keys optional).
- [x] Final `score` is consistent with documented weighting from configured weights map (MP-ART-ANALYSIS-009).
- [x] `selected` follows documented rules (for example top-K per ticker per day and minimum score) and is configurable without code deploy for routine tuning.
- [ ] POST responses `articlesScored` and `articlesSelected` match persisted outcome or errors are raised.
- [x] Chunked POSTs comply with FR9 and FR6.
- [ ] Unit tests cover breakdown shape, version field, selection edge cases, and validation failures.

## Dependencies

- Depends on: MP-ART-ANALYSIS-005
- Blocks: MP-ART-ANALYSIS-007

## Notes

Update [analysis.mdx](dev-docs/docs/mediapulse/apps/agents/analysis.mdx) and [knowledge-graph.mdx](dev-docs/docs/mediapulse/knowledge-graph/knowledge-graph.mdx) to replace legacy breakdown names (`aliasMatch`, `entityOverlap`, etc.) with canonical keys, either in this ticket or MP-ART-ANALYSIS-009 per team split. Product order breaking news, KG relation, fundamentals should appear explicitly in docs.

PRD: https://github.com/hyperjumptech/mediapulse/blob/prds/article-analysis-agent-prd.md

```json
{
  "hyper_pm_id": "01KP1FZ60DWR5GBSMXGGCCRJ7G",
  "type": "ticket",
  "parent_ids": {
    "epic": "01KP1JW1N0W9M73YC82WEWP5YY",
    "story": "01KP1JW2ZWMRP3G4YMP00KKQT4"
  },
  "depends_on": [
    "01KP1FZ60E3SX1CHQ66PQJRAE0"
  ]
}
```

### PR DESCRIPTION
**Pull request by kevinhermawan**
## Summary

Implements per-article relevance scoring for the article-analysis agent: canonical v1 `scoreBreakdown` (including `_version`), weighted `score`, Hermes-configurable selection (minimum score plus top-K per ticker per UTC day using `relevanceSelectionState` from analysis GET), and chunked `articleRelevances` POSTs after `articleEntities`. Responses aggregate `articlesScored` and `articlesSelected`.

## Related issues

Refs #216

## Important changes

- **`GET /api/v1/analysis`** now returns **`relevanceSelectionState`** (`utcDayStartIso`, `selectedCountToday`) from a single Prisma count on today’s selected rows.
- **Agent** builds heuristic breakdown, applies selection against remaining daily budget, validates chunks with `postAnalysisBodySchema.safeParse`, and posts **`article_relevances`** chunks with structured logging.

## Other changes

- Unit tests for scoring, selection, chunking, agent-data-api GET, and `run` integration.
- Updates to **analysis** and **knowledge-graph** dev-docs for canonical keys and selection rules.

## Key files to review

- `packages/shared/agent-data-api-contract/src/analysis.ts` — GET contract for `relevanceSelectionState`.
- `apps/mediapulse/agent-data-api/src/services/analysis.ts` — UTC-day count and context load.
- `apps/mediapulse/agents/article-analysis/src/run.ts` — orchestration after mention POSTs.
- `apps/mediapulse/agents/article-analysis/src/analysis-relevance-scoring.ts` — v1 breakdown and validation.
- `apps/mediapulse/agents/article-analysis/src/analysis-relevance-selection.ts` — min score, budget, tie-break.
- `apps/mediapulse/agents/article-analysis/src/analysis-relevance-post-chunks.ts` — chunked POST bodies.

## How to test

1. From repo root: `pnpm code-quality` (passes).
2. Run article-analysis against agent-data-api with Hermes config for weights / `relevanceMinScore` / `maxSelectedRelevancePerTickerPerDay`; confirm relevance rows in DB and `articlesScored` / `articlesSelected` in responses match POSTed `selected` flags.

### CODE REVIEW

File: packages/shared/agent-data-api-contract/src/analysis.ts [hunk 1]
```diff
@@ -107,11 +107,20 @@ export const analysisExistingEntitySchema = z.object({
...
+  /** Count of `articleRelevance` rows with `selected: true` and `scoredAt` on or after `utcDayStartIso`. */
+  selectedCountToday: z.number().int().nonnegative(),
+});
+
 export const getAnalysisResponseSchema = z.object({
   dataSources: z.array(analysisDataSourceSchema),
   entityTypes: z.array(analysisEntityTypeSchema),
   relationTypes: z.array(analysisRelationTypeSchema),
   existingEntities: z.array(analysisExistingEntitySchema),
+  relevanceSelectionState: analysisRelevanceSelectionStateSchema,
```
Comment (nicnocquee) [inline packages/shared/agent-data-api-contract/src/analysis.ts:133]: **Cross-cutting (#216 / GET):** Nice addition for the UTC-day budget.

**Separate gap (POST contract, same file ~L75):** `articleRelevances[].scoreBreakdown` is still `z.record(...)`. Empty `{}` and missing canonical keys are accepted. Tightening the POST schema (required `_version` + v1 keys, `superRefine`, or `.passthrough()` object) would match AC “every row has non-empty breakdown with canonical keys” for all clients.

File: apps/mediapulse/agents/article-analysis/src/analysis-relevance-post-chunks.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      relations: [],
+      articleEntities: [],
+      articleRelevances: [...window],
+    };
+    const parsed = postAnalysisBodySchema.safeParse(body);
+    if (!parsed.success) {
+      for (const issue of parsed.error.issues) {
+        parseErrors.push(issue.message);
+      }
+      continue;
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/analysis-relevance-post-chunks.ts:50]: **Silent data loss risk:** On `safeParse` failure this `continue` skips the whole window, so those `articleRelevances` never post while the run can still succeed downstream. For FR9-style chunking I’d **fail the run** (or split/retry) whenever `parseErrors` is non-empty, so we never drop scored rows quietly.

File: apps/mediapulse/agents/article-analysis/src/analysis-relevance-scoring.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+): string | null => {
+  if (row.score < 0 || row.score > 1) {
+    return "relevance score must be in [0, 1]";
+  }
+  const b = row.scoreBreakdown;
+  if (Object.keys(b).length === 0) {
+    return "scoreBreakdown must be non-empty";
+  }
+  const ver = b._version;
+  if (typeof ver !== "number" || !Number.isFinite(ver)) {
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/analysis-relevance-scoring.ts]: **#216 wording:** AC asks for `_version` as an **integer**. This allows any finite number (e.g. `1.5`). Suggest `Number.isInteger(ver)` (and `ver >= 1` if you want to mirror config schema).

File: apps/mediapulse/agents/article-analysis/src/run.ts [hunk 4]
```diff
@@ -344,9 +390,79 @@ export const run = async ({
...
+      );
+
+      const { chunks: relevanceChunks, parseErrors: relevanceParseErrors } =
+        buildArticleRelevancePostChunks(
+          input.tickerId,
+          relevanceRows,
+          cfg.postChunkArticleRelevanceBatchSize,
+        );
+
+      if (relevanceParseErrors.length > 0) {
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/run.ts:555]: Parse issues only `logger.warn`; combined with `buildArticleRelevancePostChunks` skipping bad windows, the run can finish with `success: true` and **fewer relevances persisted than built**. Consider returning failure (or treating `parseErrors.length > 0` like a hard error) so operators notice.

File: apps/mediapulse/agents/article-analysis/src/run.ts [hunk 5]
```diff
@@ -344,9 +390,79 @@ export const run = async ({
...
+    let articlesSelectedTotal = 0;
+    let relevancePostChunks = 0;
+
+    if (perSourceSignals.length > 0) {
+      const weightMap = toRelevanceWeightMapV1(cfg);
+      const relevanceDrafts = perSourceSignals.map((sig) =>
+        buildDraftRelevanceRow(sig, cfg.scoreBreakdownVersion, weightMap),
+      );
+      for (const row of relevanceDrafts) {
+        const vErr = validateRelevanceRowForPost(row, weightMap);
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/run.ts:511]: `validateRelevanceRowForPost` failures are logged but **selection and POST still run** with the same drafts. Today drafts should always pass; if that ever regresses, bad rows could ship. Consider `return { success: false, ... }` when any row fails, or filter out invalid rows before selection.

**Review by nicnocquee — COMMENTED** [review]
​

**Comment by nicnocquee** [comment]
We call `analysis.create` several times in one run (entities/relations, then mentions, then relevances). If something fails partway through, whatever already succeeded is in the DB—there’s no single rollback across the whole thing. We should document that so nobody assumes one atomic run.

**Review by nicnocquee — CHANGES_REQUESTED** [review]

File: packages/shared/agent-data-api-contract/src/analysis.ts (hunk 1 above)
Comment (kevinhermawan) [reply packages/shared/agent-data-api-contract/src/analysis.ts:133]: Implemented: `postAnalysisBodySchema.articleRelevances[].scoreBreakdown` is now a strict canonical object with required keys (`_version`, `breakingNews`, `kgRelation`, `fundamental`, `tickerSalience`, `sourceQuality`) and `_version` is constrained to integer >= 1. This enforces canonical-key presence at the contract layer for all clients.

File: apps/mediapulse/agents/article-analysis/src/analysis-relevance-post-chunks.ts (hunk 2 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/analysis-relevance-post-chunks.ts:50]: Implemented end-to-end with fail-fast behavior in `run.ts`: if `buildArticleRelevancePostChunks` reports any parse errors, the run now returns `{ success: false, ... }` with structured parse details instead of continuing. This removes the silent-drop path for scored rows.

File: apps/mediapulse/agents/article-analysis/src/analysis-relevance-scoring.ts (hunk 3 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/analysis-relevance-scoring.ts]: Implemented: `validateRelevanceRowForPost` now enforces `scoreBreakdown._version` as an integer >= 1 (`Number.isInteger` + lower bound), and I added a unit test covering non-integer rejection.

File: apps/mediapulse/agents/article-analysis/src/run.ts (hunk 4 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/run.ts:555]: Implemented: relevance parse issues are now treated as hard failures. In `run.ts`, non-empty `relevanceParseErrors` returns `success: false` with a parse error count/sample in `details`, rather than warning-only + success.

File: apps/mediapulse/agents/article-analysis/src/run.ts (hunk 5 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/run.ts:511]: Implemented: pre-selection row validation is now enforced. If any `validateRelevanceRowForPost` check fails, run exits with `success: false` and structured validation errors in `details` (instead of warning and continuing). Added run-level test coverage for this failure path.

**Review by nicnocquee — APPROVED** [review]
